WebGLRenderer: Avoid creating per-camera render states.

Oletus opened the pull request:
Now we usually only create one render state per scene. Instead of creating camera-specific render states, the view-dependent light uniforms are updated with a separate function whenever rendering. This helps avoid initMaterial calls when rendering cube maps or array cameras, improving performance dramatically.

The only exception is handling nested render() calls done from callbacks. These are sometimes used to render the same scene from a different viewpoint, for example to generate a mirror reflection. They could also potentially modify the scene in some other ways. For these reasons they can't share the same render state.

Related issues: #12883

Oletus committed:
Avoid creating per-camera render states

Now we usually only create one render state per scene. Instead of creating camera-specific render states, the view-dependent light uniforms are updated with a separate function whenever rendering. This helps avoid initMaterial calls when rendering cube maps or array cameras, improving performance dramatically.

The only exception is handling nested render() calls done from callbacks. These are sometimes used to render the same scene from a different viewpoint, for example to generate a mirror reflection. They could also potentially modify the scene in some other ways. For these reasons they can't share the same render state.

Related issues: #12883

Mugen87 commented on src/renderers/WebGLRenderer.js:
```diff
@@ -58,6 +58,11 @@ function WebGLRenderer( parameters ) {
+	// render() can be called from within a callback triggered by another render.
+	// We track this so that the nested render call gets its state isolated from the parent render call.
+
+	const currentRenderCallStack = [];
```
I guess I would call this variable `renderStateStack` since it actually holds render states.

Oletus replied:
Done! 👍

Mugen87 commented on src/renderers/WebGLRenderer.js:
```diff
@@ -1088,8 +1095,18 @@ function WebGLRenderer( parameters ) {
 		// _gl.finish();
 
+		currentRenderCallStack.pop();
+		if ( currentRenderCallStack.length > 0 ) {
```
I wonder if it's not better to hide this logic in `WebGLRenderStates` (maybe in a new method `reset()`)? In this way, you could also move the stack into this class.

Oletus replied:
`renderStates.get()` is called also in `compile()`. I don't think I'd hide this logic in `WebGLRenderStates`.

Mugen87 replied:
I see.

Mugen87 commented on src/renderers/WebGLRenderer.js:
```diff
@@ -1268,7 +1281,6 @@ function WebGLRenderer( parameters ) {
 	function renderObject( object, scene, camera, geometry, material, group ) {
 
 		object.onBeforeRender( _this, scene, camera, geometry, material, group );
-		currentRenderState = renderStates.get( scene, _currentArrayCamera || camera );
```
This is not necessary anymore since the pervious render state is now always restored and the end of `render()`, right?

Oletus replied:
Yep, that's right.

Mugen87 commented on src/renderers/webgl/WebGLRenderStates.js:
```diff
@@ -54,26 +61,26 @@ function WebGLRenderStates() {
 	let renderStates = new WeakMap();
 
-	function get( scene, camera ) {
+	function get( scene, renderCallNesting ) {
```
`renderCallNesting` sounds not descriptive enough. In some sense, it represents the depth or level of nested render calls. Maybe just `level`?

Oletus replied:
How about `renderCallDepth` or `renderCallLevel`? `level` by itself doesn't mean much.

Mugen87 replied:
Then I vote for `renderCallDepth`.

Mugen87 commented on src/renderers/webgl/WebGLRenderStates.js:
```diff
@@ -54,26 +61,26 @@ function WebGLRenderStates() {
 		} else {
 
-			if ( renderStates.get( scene ).has( camera ) === false ) {
+			if ( renderCallNesting >= renderStates.get( scene ).length ) {
```
This code section is a bit hard to read. I wonder if it's not possible to write it simpler. Couldn't you just use a `Map`?

Oletus replied:
I think it would be more confusing if a `Map` was used when the keys are indexes starting from zero.

Mugen87 commented:
> This helps avoid initMaterial calls when rendering cube maps or array cameras, improving performance dramatically.

Could you provide some performance numbers here? "improving performance dramatically" is a bit vague for my taste. Besides, the performance of `initMaterial()` calls should be already better since #19673.

Oletus commented:
> > This helps avoid initMaterial calls when rendering cube maps or array cameras, improving performance dramatically.
> 
> Could you provide some performance numbers here? _improving performance dramatically_ is a bit vague for my taste. Besides, a previous performance issue in context of `initMaterial()` was already solved via #19673.

Thanks for making me aware of #19673, I think with the effect will be less dramatic after that change. In our application we saw a more than 2x performance improvement with an earlier version of this change when rendering cubemaps, but that was without #19673. I'd still expect a fairly large benefit in our case, where we have a rather compex scene with lots of materials, with added cost from customProgramCacheKey() for a lot of materials. It could be worthwhile to take another measurement, but that will take a while.

Oletus committed:
Improve naming related to render call stack

Oletus commented:
Added a new commit with suggestions for improved naming.

One other thing: if an error is thrown from somewhere within render(), the render state will be left in the stack. Some apps might otherwise ignore the error and continue executing, in which case the stack will keep growing. Should we add a try-catch block that cleans up the stack in case of an error?

Mugen87 commented:
>  It could be worthwhile to take another measurement, but that will take a while.

Using the scene and camera to organize render states always seemed reasonable to me. The mentioned performance issues are mainly introduced of the logic in `initMaterial()`. Since the engine has mitigated this overhead starting from `r119`, it would be good to know what benefit this PR adds.

Mugen87 commented:
> Should we add a try-catch block that cleans up the stack in case of an error?

Um, not sure yet. Please give me some time to evaluate your changes in a closer manner.

Oletus commented:
I measured the difference this patch makes again, and the benefit is smaller now thanks to #19673 already fixing things. The measurement was done using Chrome's profiler. With #19673 applied but without this patch about 7.5% of CPU time is spent in initMaterial. With this patch all those calls disappear. I'd say that's still substantial enough to justify the patch, but what do you think?

Oletus commented:
It's also worth mentioning that our app probably isn't the worst case when it comes to initMaterial calls - we have an optimization to use just a single camera in our cube map generation specifically to work around the issue fixed by this PR. At its worst the overhead could be ~6x more than the 7.5% of CPU time seen in our app.

Oletus commented:
@Mugen87 any new thoughts about this? I still think the benefits are large enough especially for apps that render a lot of cube maps.

Mugen87 commented:
Sorry, I had no time yet to continue my review.

Oletus commented:
I still think this change or something like it has a lot of value - so here's a friendly ping @Mugen87 @mrdoob , any possibility of getting this merged? I could also work on an alternative version if there are any ideas how to implement this more cleanly.

Mugen87 commented on src/renderers/webgl/WebGLRenderStates.js:
```diff
@@ -54,26 +61,26 @@ function WebGLRenderStates() {
 	let renderStates = new WeakMap();
 
-	function get( scene, camera ) {
+	function get( scene, renderCallDepth ) {
```
Let's use a default parameter here:
```js
function get( scene, renderCallDepth = 0 ) {
```
This will make `renderCallDepth` optional in `WebGLRenderStates.d.ts`. And this will simplify the code in `WebGLRenderer.compile()` to `currentRenderState = renderStates.get( scene );`

Oletus replied:
Done!

Mugen87 commented:
Apart form https://github.com/mrdoob/three.js/pull/20422#discussion_r511870082, I was not able to figure out a better solution for this issue now. I think the PR is okay and an improvement compared to the previous render state handling.

Oletus committed:
Merge branch 'dev' into defer-light-transform

Oletus committed:
Use a default parameter for renderCallDepth

Oletus commented:
Addressed the comment, should be ready to merge. Thanks for the review!

Mugen87 reviewed (approved):
I suggest to merge this at the beginning of a dev cycle. Just to be on the safe side^^.

Small TODO: We should verify if the following code block is still required in its current form:

https://github.com/mrdoob/three.js/blob/956292cb4c4a90550fa5bda219b0f4a3648f5a78/src/renderers/webgl/WebGLCubeMaps.js#L43-L53

It was necessary so far to store the current render target, list and state when converting from equirect to cube map and restore them after the process. It seems this is no longer necessary for the render state (because of the new stack usage).

@Oletus Can you confirm this? 😇

Mugen87 commented:
If this is actually the case, we might want to consider the stack approach for render lists and targets, too (if possible). This would make nested renderings less error-prone.

mrdoob commented:
Sorry we didn't get to merge this in this dev cycle. Lets try next cycle!

Could you resolve the conflicts?

Oletus committed:
Merge branch 'dev' into defer-light-transform

Oletus committed:
Remove manual render state setting

Nested render calls now automatically restore the correct render state from the stack.

getRenderState/setRenderState are not part of the public API so they can be safely removed from WebGLRenderer.

Oletus commented:
@Mugen87 You're right, setting the render state in WebGLCubeMaps isn't required anymore, I pushed another commit that addresses that as well as fixing the merge conflict. I think the stack approach could make sense also for render lists, but I'd leave that to be addressed later.

The thing that I'm still most worried about with this change is that exceptions inside the renderer can cause the render state stack to grow, which could make dealing with programming errors trickier (including errors in callbacks).

Oletus commented:
@mrdoob Should be ready to be merged!

Oletus commented:
Ping @mrdoob , I think you wanted to get this merged in the beginning of this dev cycle?

mrdoob commented:
Alright! Lets give it a try 🤞

mrdoob commented:
I like this stack approach. I wish I would have came up with it 😁

Mugen87 commented:
Me too^^. 

@Oletus Cool PR!

Oletus commented:
Thanks for getting this merged! 👍
